DiamondLightSource/SynchWeb
# LIMS-121: Allow use of VMXm Gridboxes and Cartridges

**JIRA ticket**: [LIMS-121](https://jira.diamond.ac.uk/browse/LIMS-121)

**Summary**:

For VMXm to use ISPyB, we need to support their gridboxes (4 samples) and cartridges (5 samples). The expected workflow is to create multiple gridboxes with samples, and also one or more empty cartridges. Samples are then moved from the gridbox to a cartridge for loading on the beamline.

**Changes**:
- Allow creation of empty containers
- Add schematic drawings of VMXm-GridBox, VMXm-Cartridge, and also Block-4 as used on i23.
- Allow viewing of VMXm container types as "pucks" (not "plates")
- Pass in and display drawings of different container types. Define/update sample positions to match the drawings.
- Fix bug where changing container type to a new one of the same capacity wouldn't update the drawing, by adding NAME to the initial container type

**To test**:
- Add the 2 new container types to the database with:
```
insert into ContainerType (name, proposalType, capacity) values ('VMXm-Cartridge', 'mx', 5), ('VMXm-GridBox' ,'mx', 4);
```
- Create a shipment and then click Add Container
- Check puck image appears, including Mickey Mouse style ears in the middle
- Switch container type to Block-4, check square image appears and sample numbers appear
- Switch container type to VMXm-GridBox, check round image appears and sample numbers appear
- Name the grid box and add some samples to it, click Add Container, check it goes into the database
- Create another container in the same shipment, but use container type of cartridge. Check correct drawing appears.
- Dont add any samples to the cartridge, click Add Container, check no errors appear.
- Go back to the gridbox, click on the arrows to move a sample to the cartridge, check the sample moves ok.
